### ISSUE TITLE
Models failing with error - Module has no input attribute

### ISSUE THREAD

**Issue by JakeCowton** [pytorch/captum#199]
I am working with a number of models from the [torchreid](https://kaiyangzhou.github.io/deep-person-reid/) library. When I use `DeepLift` on these models, some work and some do not. For example, the [DenseNet](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/densenet.py), [MLFN](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/mlfn.py), and [MuDeep](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/mlfn.py) models work fine, but the [OSNet](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/densenet.py), [ResNetMid](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/resnetmid.py), and [ResNet-50](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/resnet.py) (and some others) model do not. (N.B. I modified the models to not use `inplace=True` for `nn.ReLU()`.)

These models that fail usually fail with an error along the lines of `'Sigmoid' object has no attribute 'input'` (though it also fails for the same reason if `ReLU` is used), however I  can't see what I need to change in this model in order for it to work with `DeepLift`.

What is different about these models that cause this error? I understand the error message, but I don't understand why the the module doesn't have an `input` attribute.

**Comment by NarineK** [comment]
Hi @JakeCowton! Thank you for the feedback! For the `ReLU'`s with `inplace=True` backward hooks do not work as expected. We need to have a special handling for it. I will look into it.

It should have worked with the  ResNet-50 and others as well. Let me try it out. I'll get back to you.

**Comment by NarineK** [comment]
Hi @JakeCowton, I looked into resnet models and it looks like ReLU is being reused in couple places in the model. Since we attach the inputs to the module if it gets reused then we wouldn't be able to map and identify the right input for that module. That's why I'd recommend to use `Relu` only once and recreate/redefine a new one if you need it. We are planning to support reuse of modules but at this point in time we need to recreate and not reuse them. 

I'm looking into `inplace=True` for ReLU's and would hopefully push the fix soon!

**Comment by akshatha-k** [comment]
@NarineK I did try implementing `DeepLift` just like @JakeCowton with the `inplace=True` but it returns 

> AttributeError: 'ReLU' object has no attribute 'input_ref'

Any suggestions on how to implement the solution you proposed with pretrained models, such as ResNets ?

**Comment by NarineK** [comment]
Hi @JakeCowton ! I just pushed the changes for supporting in-place Modules, e.g. Relus, for Deep Lift. Let me know how it works for you and if we can close this issue.
Thank you!

**Comment by NarineK** [comment]
@JakeCowton closing this since we haven't heard from you! Feel free to open a new issue if you have  any further questions.

**Comment by JakeCowton** [comment]
Hi apologies for the quietness. Christmas is a crazy time of the year!

Your PR appears to have solved the `inplace` problem (though I had solved this on my side by rewriting the models to not use `inplace`).

However I still get a `object has no attribute 'input'` for the models I mentioned in my initial issue. (e.g. [here](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/osnet.py#L217) where the `Sigmoid` function is used. What I'm trying to understand is why the [DenseNet](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/densenet.py), [MLFN](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/mlfn.py), and [MuDeep](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/mudeep.py) models work fine when they call activation functions, but  [OSNet](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/osnet.py), [ResNetMid](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/resnetmid.py), and [ResNet-50](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/models/resnet.py) (and some others) get this error when there doesn't seem to be any difference between how the data is flowing through each of the models.

I tried swapping activation functions for different ones as a sanity check and the other functions failed with the same `object has no attribute 'input'` error.
